File: src/services/audio.ts
import { invoke } from "@tauri-apps/api/core";

const STEP = 2;

export interface AudioState {
  volume: number;
  muted: boolean;
}

export const AudioService = {
  getVolume() {
    return invoke<number>("get_volume");
  },

  setVolume(volume: number) {
    return invoke<void>("set_volume", {
      volume,
    });
  },

  volumeUp(step = STEP) {
    return invoke<void>("volume_up", {
      step,
    });
  },

  volumeDown(step = STEP) {
    return invoke<void>("volume_down", {
      step,
    });
  },

  toggleMute() {
    return invoke<void>("toggle_mute");
  },

  isMuted() {
    return invoke<boolean>("is_muted");
  },
};

File: src-tauri/src/commands/audio.rs
use crate::services::audio_service::AudioService;

#[tauri::command]
pub fn get_volume() -> Result<u8, String> {
    AudioService::new()
        .map_err(|e| e.to_string())?
        .volume()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn set_volume(volume: u8) -> Result<(), String> {
    AudioService::new()
        .map_err(|e| e.to_string())?
        .set_volume(volume)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn volume_up(step: u8) -> Result<(), String> {
    AudioService::new()
        .map_err(|e| e.to_string())?
        .volume_up(step)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn volume_down(step: u8) -> Result<(), String> {
    AudioService::new()
        .map_err(|e| e.to_string())?
        .volume_down(step)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn toggle_mute() -> Result<(), String> {
    AudioService::new()
        .map_err(|e| e.to_string())?
        .toggle_mute()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn is_muted() -> Result<bool, String> {
    AudioService::new()
        .map_err(|e| e.to_string())?
        .muted()
        .map_err(|e| e.to_string())
}
